### DISCUSSION TITLE
ESP32 ESP-NOW/TCP sending files MicroPython [uasyncio]- Suggestions

### DISCUSSION THREAD

**Discussion by rarefeen** [micropython/micropython#11691] [ESP32]
I am quite new at this, does anyone know which would good protocol or method to send audio files between two esp32?
I saw quite few articles about using TCP and webserver, but did not find any on ESP-NOW, maybe because it is quite new.
I try to do it using uasyncio startserver to make server and send files using various methods [ so far failed miserably].
so I am looking for methods and opinions whether anyone has done it with uasyncio start server, then connect with client to send audio file. or used ESP-NOW protocol to send files.
Thank you for any advices.
[forgive me if the question is too basic or I made a mistake in my writing.]

**Comment by bixb922** [comment]
Perhaps this could help on the ESP32 side; https://github.com/robert-hh/FTP-Server-for-ESP8266-ESP32-and-PYBD
Installing and importing uftp.py starts a FTP server on the ESP32. On the PC there are many clients available, from command line curl to drag and drop. You'll need to start wifi.

**Reply by rarefeen** [reply]
@bixb922 Thanks for it. I was actually looking for esp32 to esp32 communication [which will be a function of button ]. the purpose is to send audio from one to another esp32.
EDIT:
I saw the functions which has send or receive functions. not sure if i can use two esp32 as in the readme, there is no indications of it. also, I wonder if it has high error rates as they gave some uncertainties in the readme.

**Reply by bixb922** [reply]
There are FTP clients for Micropython. If not FTP, I'd use urequests and Microdot or tinyweb for file transfer over http (instead of coding that at socket level)

**Reply by rarefeen** [reply]
can i use tinyweb/microdot in one esp32 to make server, then use another two esp32 [without tinyweb installed in it] to send file?
EDIT:
what i meant is:
esp32 i => act as server [has Microdot or tinyweb  server]
esp32 ii => is client [does not have Microdot or tinyweb  server]. but it should send/receive files.
esp32 iii => is client [does not have Microdot or tinyweb  server]. but it should send/receive files.

i want to communicate between esp32 ii and iii to send data in between them through i. I was hoping I could find some solution with ESP NOW so that I donot have to use 3 esp32. it seems it is not possible maybe.

**Reply by bixb922** [reply]
Yes, and completing the description: To have a ESP32  initiate a communication, use urequests. To have the ESP32 be the passive side waiting for a communication to start, use Microdot or tinyweb. Both can send and receive files. The difference is who can initiate the communication. 
A ESP32 can have both urequests and Microdot or tinyweb active and in use.

**Reply by rarefeen** [reply]
Thank you so much for the explanation. I will check if it fits my needs and the memory i have left in esp32. btw, do you have any examples? I would like to see as reference to make myself understand a little bit faster.

**Reply by peterhinch** [reply]
@rarefeen It's worth pointing out that the above good suggestions are based on sending an audio file from one board to another. Once the file has been received, it can be played.

This is not the same as realtime audio, for example in a baby alarm, where one device produces a continuous stream of audio which must be played as it occurs. This is a much more difficult problem.

**Reply by rarefeen** [reply]
@peterhinch Thank you for confirming. at this point I just want to send the audio file, not doing realtime playing the audio. Since I am running aioespnow too do to other activities. I wonder if microdot/tinydot it will hinder the other work. I did not had the time to check it yet. If you have any examples I could really appreciate it [as it is quite time consuming to read through everything to make it. If example works then I can read it and change as per my needs.]

**Reply by peterhinch** [reply]
Sorry but I have no experience of espnow (yet - it's on my list).

**Reply by glenn20** [reply]
ESPNow works fine with asyncio (you may need to `mip install aioespnow`), however it is just a protocol for sending and receiving messages (max 250 bytes), akin to udp and does not provide a stream protocol layer. To reliably send audio files, you would need to implement some simple protocol to ensure reliable delivery of the entire file contents. It would be simpler to use microdot and urequests (I use that for large file transfers rather than espnow - eg. OTA images).

(I have been working on some simple pseudo-protocol layers for service discovery and service end points over espnow - including a stream-like interface, but I don't know when I'll be able to get back to that work).

**Reply by rarefeen** [reply]
@glenn20  @peterhinch Thank you for the advice. I will try them. 
@peterhinch, I am running espnow for different reason. but I will use microdot/tinydot only for the audio/img file send/recv. so I was looking for example which contains only using microdot/tinydot for files.
@glenn20 I won't be using espnow protocol for sending audio. I want to use microdot-urequest to send/receive data along with espnow. Is this possible?

**Reply by glenn20** [reply]
> @peterhinch, I am running espnow for different reason. but I will use microdot/tinydot only for the audio/img file send/recv. so I was looking for example which contains only using microdot/tinydot for files.
> @glenn20 I won't be using espnow protocol for sending audio. I want to use microdot-urequest to send/receive data along with espnow. Is this possible?

Yes, you can use espnow and wifi together, although there are a few traps for the unwary. Check out https://docs.micropython.org/en/latest/library/espnow.html#espnow-and-wifi-operation.

However, if you are already using wifi to send audio data, you may just be complicating your life to send other data with espnow, though I can imagine there can be good reasons to do so.

**Reply by peterhinch** [reply]
@glenn20 I'll try `aioespnow` in the next few days. Do all ESP32 variants support it? (It might be good to clarify this in the doc).

**Reply by glenn20** [reply]
> @glenn20 I'll try `aioespnow` in the next few days. Do all ESP32 variants support it? (It might be good to clarify this in the doc).

@peterhinch I believe so. I have tested on the S2 and S3, but I haven't tested the C3.

**Reply by glenn20** [reply]
And yes. You are right about the docs.

**Reply by peterhinch** [reply]
@glenn20 Sadly `aioespnow` is not available in S3 daily builds:
```
MicroPython v1.20.0-162-gd080d427e on 2023-06-06; FeatherS3 with ESP32-S3
Type "help()" for more information.
>>> import espnow
>>> import aioespnow
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: no module named 'aioespnow'
>>> 
```

**Reply by glenn20** [reply]
Damien requested that I move it to micropython_lib, which has been done, but I had to remove aioespnow from the manifest.py until the micropython builds migrate to the latest micropython-lib. Until that is all sorted you have to `mip install aioespnow`.

**Reply by peterhinch** [reply]
I guess this is life at the bleeding edge, but I get `ValueError: invalid buffer length` on initialisation:
```python
import network
import aioespnow
import uasyncio as asyncio
#from mac import mac1 as peer

network.WLAN(network.STA_IF).active(True)
e = aioespnow.AIOESPNow()  # Returns AIOESPNow enhanced with async support
#print(e.config("rxbuf"))
e.active(True)
e.add_peer(b"\x70\x04\x1d\xad\x90")  # Fails here
```
Also `print(e.config("rxbuf"))` fails with `ValueError: unknown config param`.
```
MicroPython v1.20.0-162-gd080d427e on 2023-06-06; FeatherS3 with ESP32-S3
```
This isn't confined to `aioespnow`. After a soft reset I tried:
```python
>>> import espnow
>>> import network
>>> network.WLAN(network.STA_IF).active(True)
True
>>> e = espnow.ESPNow()
>>> e.active(True)
True
>>> e.add_peer(b"\x70\x04\x1d\xad\x90")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid buffer length
>>> 
```
Also on ESP32-S2:
```python
MPY: soft reboot
MicroPython v1.20.0-162-gd080d427e on 2023-06-06; FeatherS2 with ESP32-S2
Type "help()" for more information.
>>> 
>>> import network, espnow
>>> network.WLAN(network.STA_IF).active(True)
True
>>> e = espnow.ESPNow()
>>> e.active(True)
True
>>> e.add_peer(b"\x70\x04\x1d\xad\x90")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid buffer length
>>> 
```

**Reply by glenn20** [reply]
> I guess this is life at the bleeding edge, but I get `ValueError: invalid buffer length` on initialisation:
> ```python
> import network
> import aioespnow
> import uasyncio as asyncio
> #from mac import mac1 as peer
> 
> network.WLAN(network.STA_IF).active(True)
> e = aioespnow.AIOESPNow()  # Returns AIOESPNow enhanced with async support
> #print(e.config("rxbuf"))
> e.active(True)
> e.add_peer(b"\x70\x04\x1d\xad\x90")  # Fails here
> ```
> Also `print(e.config("rxbuf"))` fails with `ValueError: unknown config param`.
> ```
> MicroPython v1.20.0-162-gd080d427e on 2023-06-06; FeatherS3 with ESP32-S3
> ```
> This isn't confined to `aioespnow`. After a soft reset I tried:
> ```python
> >>> import espnow
> >>> import network
> >>> network.WLAN(network.STA_IF).active(True)
> True
> >>> e = espnow.ESPNow()
> >>> e.active(True)
> True
> >>> e.add_peer(b"\x70\x04\x1d\xad\x90")
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> ValueError: invalid buffer length
> >>> 
> ```
> Also on ESP32-S2:
> ```python
> MPY: soft reboot
> MicroPython v1.20.0-162-gd080d427e on 2023-06-06; FeatherS2 with ESP32-S2
> Type "help()" for more information.
> >>> 
> >>> import network, espnow
> >>> network.WLAN(network.STA_IF).active(True)
> True
> >>> e = espnow.ESPNow()
> >>> e.active(True)
> True
> >>> e.add_peer(b"\x70\x04\x1d\xad\x90")
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> ValueError: invalid buffer length
> >>> 
> ```

That's because the Mac address is only 5 bytes long ;).

I'll look into the issue with getting the value of the rxbuf parameter.

**Reply by rarefeen** [reply]
@peterhinch not sure if this will help of not.
i use this to get mac and add in pair. it's kind of easier for me. to get rid of ValueError: invalid buffer length. I suffered because of this several times and try to make it simpler.
on device a:
`wlan_mac = ubinascii.hexlify(sta_if.config('mac')).decode()`
on device b to pair with device a:
`sp_now.add_peer(ubinascii.unhexlify(wlan_mac))`

**Reply by glenn20** [reply]
> Also `print(e.config("rxbuf"))` fails with `ValueError: unknown config param`.

@peterhinch Thanks for reporting this. The docs are out of date: replace "rxbuf" with "buffer" (suggests this parameter is not in wide use ;-). I'll post a fix. (In earlier versions there was a tx and an rx buffer, but the txbuffer was unnecessary - and using "buffer" saved one more QSTR slot ;-). I only change this value when I want to run some high-throughput applications.

UPDATE: to clarify, on esp32 the "buffer" config() option will be renamed to "rxbuf" (which matches the docs and esp8266). See PR #11747.

**Reply by glenn20** [reply]
> `wlan_mac = ubinascii.hexlify(sta_if.config('mac')).decode()`

Micropython has added the `.hex()` method to `bytes`, so you can simplify this to:
```py
wlan_mac = sta_if.config('mac').hex()
sp_now.add_peer(bytes.fromhex(wlan_mac))
```
~~However, you still need unhexlify for converting the other way.~~ (Whoops - I forgot `fromhex()` had been implemented as well)

**Reply by peterhinch** [reply]
> That's because the Mac address is only 5 bytes long ;)

So it is. Life at the geriatric edge :) Now working.

**Reply by glenn20** [reply]
> So it is. Life at the geriatric edge :) Now working.

Ha. I'm also right there....

**Reply by rarefeen** [reply]
@glenn20 Thank you, is it on v1.20 or 1.19 also supports?

**Reply by glenn20** [reply]
I believe it is in 1.20, but not 1.19.

**Reply by rarefeen** [reply]
@glenn20 @peterhinch sorry if this is a dumb question,
But is there any way to change the micropython firmware [for example, custom install few libraries and make a main.py and other files ]. then make it into firmware file?
so I can use it[same custom firmware] on multiple esp32 devices? If it is possible, how to do it, is there any guideline?
for example, if I install microdot on micropython V1.2, then I want to make it into firmware so that I do not need to install it on other esp32's.

**Reply by peterhinch** [reply]
See [this doc](http://docs.micropython.org/en/latest/reference/manifest.html#manifest) which explains frozen bytecode. Another approach is to use standard firmware but use `mpremote` to install the modules you need. `mpremote` can be scripted, so you would only need to run the script against each target to adapt it.

**Reply by rarefeen** [reply]
@peterhinch Thank you, I will look into it now.
for more clarification, What I ultimately wanted is, I have several esp32, I wanted to make a .bin file [firmware], 
meaning I want to upload codes into esp32 then just export the firmware including the files from esp32 in a .bin.
